File: server/internal/models/nodereg/gate_upstream_test.go
package nodereg

import (
	"testing"

	"github.com/cocofhu/approving/internal/models"
)

func TestGatePrimaryUpstreamNodeID(t *testing.T) {
	// page preferred over earlier structured ref
	gate := &models.Node{
		ID: "gate", Type: "human_gate",
		Config: map[string]any{
			"body_template": "see {{nodes.research.outputs.research}} and {{nodes.visual.outputs.page}}",
		},
	}
	if got := GatePrimaryUpstreamNodeID(gate); got != "visual" {
		t.Fatalf("page preferred: got %q", got)
	}

	// no page → first structured
	gate2 := &models.Node{
		ID: "gate", Type: "human_gate",
		Config: map[string]any{"body_template": "{{nodes.plan.outputs.plan}}"},
	}
	if got := GatePrimaryUpstreamNodeID(gate2); got != "plan" {
		t.Fatalf("first structured: got %q", got)
	}

	// no refs → empty
	gate3 := &models.Node{ID: "gate", Type: "human_gate", Config: map[string]any{"body_template": "plain"}}
	if got := GatePrimaryUpstreamNodeID(gate3); got != "" {
		t.Fatalf("no refs: got %q", got)
	}
	if got := GatePrimaryUpstreamNodeID(nil); got != "" {
		t.Fatalf("nil node: got %q", got)
	}
}

func TestGatePrimaryProducts_multiAndArtifact(t *testing.T) {
	gate := &models.Node{
		ID: "gate", Type: "human_gate",
		Config: map[string]any{
			"body_template": `{{nodes.react.outputs.clarified_requirement}}
{{artifact("notes.md")}}
and {{nodes.visual.outputs.page}}`,
		},
	}
	got := GatePrimaryProducts(gate, []string{"clarified_requirement.json", "extra.md"})
	if len(got) != 3 {
		t.Fatalf("want 3 products, got %d: %+v", len(got), got)
	}
	byName := map[string]bool{}
	for _, p := range got {
		byName[p.Name] = true
	}
	for _, w := range []string{"clarified_requirement.json", "notes.md", "page.html"} {
		if !byName[w] {
			t.Fatalf("missing %q in %+v", w, got)
		}
	}
	// produces-only names must not invent editables
	if GateAllowsArtifact(gate, "extra.md", []string{"extra.md"}) {
		t.Fatal("produces-only artifact must not be editable")
	}
	if !GateAllowsArtifact(gate, "notes.md", nil) {
		t.Fatal("artifact() ref must be editable")
	}
}

func TestGatePrimaryProducts_proposalSelect(t *testing.T) {
	gate := &models.Node{ID: "sel", Type: "proposal_select", Config: map[string]any{}}
	got := GatePrimaryProducts(gate, nil)
	if len(got) != 1 || got[0].Name != "proposals.json" {
		t.Fatalf("proposal_select default: %+v", got)
	}
	gate.Config["from"] = "alts.json"
	got = GatePrimaryProducts(gate, nil)
	if len(got) != 1 || got[0].Name != "alts.json" {
		t.Fatalf("proposal_select from: %+v", got)
	}
}

func TestInferArtifactKind_imageAndReadonly(t *testing.T) {
	cases := []struct {
		name     string
		wantKind string
		readonly bool
	}{
		{"shot.png", "image", true},
		{"photo.JPG", "image", true},
		{"a.jpeg", "image", true},
		{"b.webp", "image", true},
		{"c.gif", "image", true},
		{"page.html", "html", false},
		{"notes.md", "markdown", false},
		{"research.json", "json", false},
		{"plain.txt", "text", false},
	}
	for _, tc := range cases {
		got := InferArtifactKind(tc.name)
		if got != tc.wantKind {
			t.Fatalf("%s: kind want %q got %q", tc.name, tc.wantKind, got)
		}
		if IsReadonlyArtifactKind(got) != tc.readonly {
			t.Fatalf("%s: readonly want %v", tc.name, tc.readonly)
		}
	}
	if !IsNonTextPrimary("shot.png", "") {
		t.Fatal("suffix image must be non-text")
	}
	if !IsNonTextPrimary("odd.bin", "image") {
		t.Fatal("store kind image must win")
	}
	if IsNonTextPrimary("notes.md", "markdown") {
		t.Fatal("markdown must remain editable")
	}
}

func TestGatePrimaryProducts_imageReadonly(t *testing.T) {
	gate := &models.Node{
		ID: "gate", Type: "human_gate",
		Config: map[string]any{
			"body_template": `{{artifact("screenshot.png")}} {{nodes.visual.outputs.page}}`,
		},
	}
	got := GatePrimaryProducts(gate, nil)
	byName := map[string]GatePrimaryProduct{}
	for _, p := range got {
		byName[p.Name] = p
	}
	img, ok := byName["screenshot.png"]
	if !ok || img.Kind != "image" || !img.Readonly {
		t.Fatalf("screenshot.png: %+v", img)
	}
	page, ok := byName["page.html"]
	if !ok || page.Kind != "html" || page.Readonly {
		t.Fatalf("page.html: %+v", page)
	}
}

File: web/src/lib/gateUpstream.test.ts
import { describe, expect, it } from 'vitest'
import {
  listPrimaryProducts,
  pickProductRef,
  resolveUpstreamOutputs,
  reviewingUpstreamN,
} from './gateUpstream'

describe('pickProductRef', () => {
  it('prefers page over an earlier structured ref', () => {
    const tmpl =
      'see {{nodes.research.outputs.research}} then {{nodes.visual.outputs.page}}'
    expect(pickProductRef(tmpl)).toEqual({ nodeId: 'visual', key: 'page' })
  })

  it('falls back to the first structured ref when no page', () => {
    expect(pickProductRef('{{nodes.plan.outputs.plan}}')).toEqual({
      nodeId: 'plan',
      key: 'plan',
    })
  })

  it('returns null when template has no node output refs', () => {
    expect(pickProductRef('plain markdown')).toBeNull()
  })
})

describe('listPrimaryProducts', () => {
  it('collects outputs refs and artifact() refs', () => {
    const tmpl = `{{nodes.react.outputs.clarified_requirement}}
{{artifact("notes.md")}}
{{nodes.visual.outputs.page}}`
    const got = listPrimaryProducts(tmpl)
    expect(got.map((p) => p.name)).toEqual([
      'clarified_requirement.json',
      'page.html',
      'notes.md',
    ])
  })

  it('falls back to proposals.json for proposal_select', () => {
    expect(listPrimaryProducts('', { isProposalSelect: true })).toEqual([
      { name: 'proposals.json', outputKey: 'proposals', kind: 'json', readonly: false },
    ])
  })

  it('marks image artifacts as readonly', () => {
    const got = listPrimaryProducts('{{artifact("shot.png")}} {{nodes.visual.outputs.page}}')
    expect(got).toEqual([
      { name: 'page.html', nodeId: 'visual', outputKey: 'page', kind: 'html', readonly: false },
      { name: 'shot.png', outputKey: undefined, kind: 'image', readonly: true },
    ])
  })
})

describe('listExcludedProduces', () => {
  it('lists produces-only names not in the template whitelist', async () => {
    const { listExcludedProduces } = await import('./gateUpstream')
    const excluded = listExcludedProduces(
      '{{nodes.test.outputs.test_result}} {{artifact("shot.png")}}',
      [
        {
          id: 'test',
          config: { produces: 'test_result.json,shot.png,extra.md' },
        },
      ],
    )
    expect(excluded).toEqual(['extra.md'])
  })

  it('includes produces-only from nodes not referenced by primary nodeId', async () => {
    const { listExcludedProduces } = await import('./gateUpstream')
    // Template mixes outputs + artifact(); visual has page.html but research
    // produces-only node_complete.json must still appear in the excluded hint.
    const excluded = listExcludedProduces(
      '{{nodes.visual.outputs.page}} {{artifact("shot.png")}}',
      [
        {
          id: 'visual',
          config: { produces: 'page.html,shot.png' },
        },
        {
          id: 'research',
          config: { produces: 'research.json,node_complete.json' },
        },
      ],
    )
    expect(excluded).toEqual(['research.json', 'node_complete.json'])
  })
})

describe('resolveUpstreamOutputs', () => {
  const execsByNode = {
    visual: [
      { iteration: 1, status: 'completed', outputs: { page: '<html>v1</html>' } },
      { iteration: 2, status: 'failed', outputs: { page: '<html>fail</html>' } },
      { iteration: 3, status: 'completed', outputs: { page: '<html>v3</html>' } },
    ],
  }

  it('binds visual iter=3 when gate iter=2 and pointer points at 3', () => {
    const r = resolveUpstreamOutputs({
      productNodeId: 'visual',
      execsByNode,
      upstreamNodeId: 'visual',
      upstreamIteration: 3,
      gateIteration: 2,
      pending: true,
    })
    expect(r.usedPointer).toBe(true)
    expect(r.pointerMiss).toBe(false)
    expect(r.selectedIteration).toBe(3)
    expect(r.outputs?.page).toBe('<html>v3</html>')
  })

  it('marks pointerMiss without falling back to equals heuristic', () => {
    const r = resolveUpstreamOutputs({
      productNodeId: 'visual',
      execsByNode,
      upstreamNodeId: 'visual',
      upstreamIteration: 9,
      gateIteration: 2,
      pending: true,
    })
    expect(r.usedPointer).toBe(true)
    expect(r.pointerMiss).toBe(true)
    expect(r.outputs).toBeNull()
    expect(r.selectedIteration).toBe(9)
    // Banner N stays on pointer; UI must show a secondary fallback hint when
    // loadProduct then reads the run artifact store.
    expect(reviewingUpstreamN({
      upstreamIteration: 9,
      selectedIteration: r.selectedIteration,
    })).toBe(9)
  })

  it('legacy pending gate picks max completed (skips failed retry)', () => {
    const r = resolveUpstreamOutputs({
      productNodeId: 'visual',
      execsByNode,
      gateIteration: 2,
      pending: true,
    })
    expect(r.usedPointer).toBe(false)
    expect(r.selectedIteration).toBe(3)
    expect(r.outputs?.page).toBe('<html>v3</html>')
  })

  it('legacy resolved gate picks nearest iteration ≤ gate.iteration', () => {
    const r = resolveUpstreamOutputs({
      productNodeId: 'visual',
      execsByNode,
      gateIteration: 2,
      pending: false,
    })
    expect(r.usedPointer).toBe(false)
    expect(r.selectedIteration).toBe(2)
    expect(r.outputs?.page).toBe('<html>fail</html>')
  })
})

describe('reviewingUpstreamN', () => {
  it('uses pointer iteration when present', () => {
    expect(
      reviewingUpstreamN({ upstreamIteration: 3, selectedIteration: 1 }),
    ).toBe(3)
  })

  it('falls back to selected iteration for legacy gates', () => {
    expect(reviewingUpstreamN({ selectedIteration: 2 })).toBe(2)
  })
})
